### ISSUE TITLE
Termination problem: copatterns and without-K

### ISSUE THREAD

**Issue by sattlerc** [agda/agda#3081]
On master.

```agda
{-# OPTIONS --without-K #-}
module Termination where

postulate
  X : Set
  P : (x : X) → Set
  any : {X : Set} → X

record W : Set where
  field
    x : X
    p : P x

record Category : Set₁ where
  field
    z : W
    s : W → W

{- Termination checking failed for the following functions:
     functor
   Problematic calls:
     W.x (functor .Category.s u)
     W.x (functor .Category.z)
-}
functor : Category
functor .Category.z .W.x = any
functor .Category.z .W.p = any
functor .Category.s u .W.x = any
functor .Category.s u .W.p = any
```
The issue disappears if --without-K is removed.

**Comment by andreasabel** [comment]
This might related to #1680, see also #2890.
The `--without-K` termination checker is more/too conservative; it does not like result types of clauses like `X` which are not related to eliminated type `Category`.

**Comment by UlfNorell** [comment]
Something is clearly wonky with the `--without-K` termination checker. Here's a simplified example:
```agda
{-# OPTIONS --without-K #-}
module _ where

open import Agda.Builtin.Nat

record R : Set where
  field
    a : Nat
    b : Nat

record S : Set where
  field
    r : R  -- works if we give r and h the same number of arguments
    h : Nat → R

open R
open S

f : S
f .r   .a = 0
f .r   .b = f .r   .a
f .h n .a = n
f .h n .b = f .h n .a
```

**Comment by guillaumebrunerie** [comment]
The following is an even more simplified example:

```agda
{-# OPTIONS --without-K #-}
open import Agda.Builtin.Nat

record R : Set where
  field
   f : Nat → Nat
open R

r : R
r .f zero = zero
r .f (suc n) = r .f n
```

**Comment by andreasabel** [comment]
The OP and Ulf's variant are fixed by PR #4424.  
Guillaume's simplification is a duplicate of #1680.
